src/github_runner_image_builder/builder.py

```diff
@@ -0,0 +1,635 @@
+        logger.info("modprobe nbd out: %s", output)
+    except subprocess.CalledProcessError as exc:
+        logger.exception(
+            "Error enabling network block device, cmd: %s, code: %s, err: %s",
+            exc.cmd,
+            exc.returncode,
+            exc.output,
+        )
+        raise NetworkBlockDeviceError from exc
+
+
+def build_image(arch: Arch, base_image: BaseImage) -> None:
+    """Build and save the image locally.
+
+    Args:
+        arch: The CPU architecture to build the image for.
+        base_image: The ubuntu image to use as build base.
+
+    Raises:
+        BuildImageError: If there was an error building the image.
+    """
+    IMAGE_MOUNT_DIR.mkdir(parents=True, exist_ok=True)
+    try:
+        logger.info("Cleaning build state.")
+        _clean_build_state()
+        logger.info("Downloading base image.")
+        base_image_path = _download_base_image(arch=arch, base_image=base_image)
+        logger.info("Resizing base image.")
+        _resize_image(image_path=base_image_path)
+        logger.info("Mounting network block device.")
+        _mount_image_to_network_block_device(image_path=base_image_path)
+        logger.info("Replacing resolv.conf.")
+        _replace_mounted_resolv_conf()
+        logger.info("Resizing partitions.")
+        _resize_mount_partitions()
+        logger.info("Installing YQ from source.")
+        _install_yq()
+    except ImageBuilderBaseError as exc:
+        raise BuildImageError from exc
+
+    try:
+        logger.info("Setting up chroot environment.")
+        with ChrootContextManager(IMAGE_MOUNT_DIR):
+            # operator_libs_linux apt package uses dpkg -l and that does not work well with
+            # chroot env, hence use subprocess run.
+            output = subprocess.check_output(
+                ["/usr/bin/apt-get", "update", "-y"],
+                timeout=60 * 10,
+                env={"DEBIAN_FRONTEND": "noninteractive"},
+            )  # nosec: B603
+            logger.info("apt-get update out: %s", output)
+            output = subprocess.check_output(  # nosec: B603
+                [
+                    "/usr/bin/apt-get",
+                    "install",
+                    "-y",
+                    "--no-install-recommends",
+                    *IMAGE_DEFAULT_APT_PACKAGES,
+                ],
+                timeout=60 * 20,
+                env={"DEBIAN_FRONTEND": "noninteractive"},
+            )
+            logger.info("apt-get install out: %s", output)
+            _disable_unattended_upgrades()
+            _configure_system_users()
+            _install_yarn()
+    except ChrootBaseError as exc:
+        raise BuildImageError from exc
+    finally:
+        _clean_build_state()
+
+    try:
+        logger.info("Compressing image.")
+        _compress_image(base_image_path)
+    except ImageBuilderBaseError as exc:
+        raise BuildImageError from exc
```

Comment:
I think `ImageBuilderBaseError` can inherit from `BuildImageError` which would avoid nesting the error